auth/src/krb.cpp

```diff
@@ -0,0 +1,942 @@
+        }
+        return SUCCESS();
+    } // krb_receive_token
+} // anonymous namespace
+
+namespace irods
+{
+    class kerberos_authentication : public irods::experimental::auth::authentication_base {
+    public:
+        kerberos_authentication()
+        {
+            add_operation(AUTH_CLIENT_AUTH_REQUEST,  OPERATION(rcComm_t, kerberos_auth_client_request));
+            add_operation(AUTH_ESTABLISH_CONTEXT,    OPERATION(rcComm_t, kerberos_auth_establish_context));
+            add_operation(AUTH_CLIENT_AUTH_RESPONSE, OPERATION(rcComm_t, kerberos_auth_client_response));
+#ifdef RODS_SERVER
+            add_operation(AUTH_AGENT_AUTH_REQUEST,   OPERATION(rsComm_t, kerberos_auth_agent_request));
+            add_operation(AUTH_AGENT_AUTH_RESPONSE,  OPERATION(rsComm_t, kerberos_auth_agent_response));
+#endif
+        } // ctor
+
+    private:
+        json auth_client_start(rcComm_t& comm, const json& req)
+        {
+            json resp{req};
+            resp["user_name"] = comm.proxyUser.userName;
+            resp["zone_name"] = comm.proxyUser.rodsZone;
+            resp["next_operation"] = AUTH_CLIENT_AUTH_REQUEST;
+
+            return resp;
+        } // auth_client_start
+
+        json kerberos_auth_client_request(rcComm_t& comm, const json& req)
+        {
+            // Setup the credentials
+            const auto service_name = krb_setup_creds(comm.rError, true, "");
+
+            json svr_req{req};
+            svr_req[irods::AUTH_USER_KEY] = service_name;
+            svr_req["next_operation"] = AUTH_AGENT_AUTH_REQUEST;
+
+            auto resp = request(comm, svr_req);
+
+            if(auto [e, n] = verify(resp, {"request_result"}); !e) {
+                THROW(SYS_INVALID_INPUT_PARAM,
+                      fmt::format("[{}:{}] missing [{}] in response", __func__, __LINE__, n));
+            }
+
+            const auto result = resp.at("request_result").get_ref<const std::string&>();
+            // TODO: request_result contains the service name? Is that the password??
+            if (const auto ec = obfSavePw(0, 0, 0, result.data()); ec < 0) {
+                THROW(ec, fmt::format(
+                      "failed to save obfuscated password while authenticating user [{}]",
+                      service_name));
+            }
+
+            resp["next_operation"] = AUTH_ESTABLISH_CONTEXT;
+
+            return resp;
+        } // kerberos_auth_client_request
+
+        json kerberos_auth_establish_context(rcComm_t& _comm, const json& req)
+        {
+            if (auto [e, n] = verify(req, {"request_result"}); !e) {
+                THROW(SYS_INVALID_INPUT_PARAM,
+                      fmt::format("[{}:{}] missing [{}] in request", __func__, __LINE__, n));
+            }
+
+            json resp{req};
+
+            ikrb_rErrorPtr = _comm.rError;
+
+            std::string service_principal;
+            if (const char* irods_server_dn = getenv("irodsServerDn")) {
```

Comment:
Consider separating the boolean check from the assignment. While this works, it could confuse readers unfamiliar with what is happening.
```c++
if (init-op; conditional) {
```

Reply:
Will do

Reply:
Done